tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala

```diff
@@ -218,45 +195,46 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
-        if (arraysEnabled) {
-          val (setExpr, setT, _) = cellCache(setId).head
-          val (elemExpr, elemT, _) = cellCache(elemId).head
-
-          // inPred can either represent an array update, i.e. a "store", or an array access, i.e. a "select", as the
-          // the current implementation is adhering to the interface made for the original encoding.
-          // TODO: Should we split this method and abandon the concept of an in-relation?
-          inStored.get(setId, elemId) match {
-            case None =>
-              val name = setExpr.toString + "_" + s"in_${elemT.signature}${elemId}_${setT.signature}$setId"
-              logWriter.flush() // flush the SMT log
-              throw new IllegalStateException(s"SMT $id: The Array constant $name is missing from the SMT context")
-
-            // Array "store"
-            case Some(list) if list.head._1 == false =>
-              val storedExpr = if (propagatedArraySelect._1) propagatedArraySelect._2 else z3context.mkTrue()
-              propagatedArraySelect = propagatedArraySelect.copy(_1 = false)
-              val store = z3context.mkStore(
-                  setExpr.asInstanceOf[Expr[ArraySort[Sort, BoolSort]]], elemExpr.asInstanceOf[Expr[Sort]],
-                  storedExpr.asInstanceOf[Expr[BoolSort]])
-              val eqStore = z3context.mkEq(const, store)
-              // inStored records that the array was set in the current level
-              inStored += ((setId, elemId) -> ((true, level) :: inStored(setId, elemId)))
-              // cellCache records the current SSA array representing the set being handled
-              cellCache += (setId -> ((const, setT, level) :: cellCache(setId)))
-              eqStore.asInstanceOf[ExprSort]
-
-            // Array "select"
-            case Some(list) if list.head._1 == true =>
-              z3context
-                .mkSelect(setExpr.asInstanceOf[ArrayExpr[Sort, BoolSort]], elemExpr.asInstanceOf[Expr[Sort]])
-                .asInstanceOf[ExprSort]
-          }
-        } else {
-          const
-        }
+        const
     }
   }
 
+  private def mkSelect(setId: Int, elemId: Int): ExprSort = {
+    val set = cellCache(setId).head._1
+    val elem = cellCache(elemId).head._1
+
+    z3context
+      .mkSelect(set.asInstanceOf[ArrayExpr[Sort, BoolSort]], elem.asInstanceOf[Expr[Sort]])
+      .asInstanceOf[ExprSort]
+  }
+
+  private def mkStore(setId: Int, elemId: Int): ExprSort = {
+    val (set, setT, _) = cellCache(setId).head
+    val (elem, elemT, _) = cellCache(elemId).head
+
+    val edgeName = s"in_${elemT.signature}${elemId}_${setT.signature}$setId"
+    val updatedSetName = set.toString + "_" + edgeName
+    val setSort = getOrMkCellSort(setT)
+    log(s";; declare edge inclusion $edgeName")
+    log(s"(declare-const $updatedSetName $setSort)")
+    val updatedSet = z3context.mkConst(updatedSetName, setSort)
+    cellCache += (setId -> ((updatedSet, setT, level) :: cellCache(setId)))
```

Comment:
That's where another cache would make things more obvious :)

Reply:
As stated in a previous comment, I'm not sure `cellCache` can work without storing a list. If this is the case I'm not sure if it would make sense to have another cache. We could split `cellCache` and have one cache hold the head of the list while the other holds the tail, but I don't see the benefit.

Reply:
Yes, I see what you mean now. I am thinking about the non-array encoding. Since it is a list, there should be no big memory penalty for storing the list instead of a single element.